Source/Core/DolphinQt2/FIFOWindow.cpp

```diff
@@ -0,0 +1,359 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "DolphinQt2/FIFOWindow.h"
+
+#include <QCheckBox>
+#include <QDialogButtonBox>
+#include <QFileDialog>
+#include <QGridLayout>
+#include <QGroupBox>
+#include <QLabel>
+#include <QMessageBox>
+#include <QPushButton>
+#include <QSpinBox>
+
+#include "Core/FifoPlayer/FifoDataFile.h"
+#include "Core/FifoPlayer/FifoPlaybackAnalyzer.h"
+#include "Core/FifoPlayer/FifoPlayer.h"
+#include "Core/FifoPlayer/FifoRecorder.h"
+
+#include "DolphinQt2/QtUtils/QueueOnObject.h"
+#include "DolphinQt2/Settings.h"
+
+FIFOWindow::FIFOWindow(QWidget* parent) : QDialog(parent)
+{
+  setWindowTitle(tr("FIFO Player"));
+
+  CreateWidgets();
+  ConnectWidgets();
+
+  UpdateInfo();
+
+  connect(this, &FIFOWindow::EmulationStarted, this, &FIFOWindow::OnEmulationStarted);
+  connect(this, &FIFOWindow::EmulationStopped, this, &FIFOWindow::OnEmulationStopped);
+  connect(this, &FIFOWindow::EmulationStopped, this, &FIFOWindow::UpdateInfo);
+
+  connect(&Settings::Instance(), &Settings::FIFORecordingDone, this, &FIFOWindow::OnRecordingDone);
+  connect(&Settings::Instance(), &Settings::FIFOLogLoaded, this, &FIFOWindow::OnFIFOLoaded);
+  connect(&Settings::Instance(), &Settings::FIFOFrameWritten, this, &FIFOWindow::UpdateInfo);
+
+  FifoPlayer::GetInstance().SetFileLoadedCallback([] {
+    QueueOnObject(&Settings::Instance(), [] { Settings::Instance().TriggerFIFOLogLoaded(); });
+  });
+  FifoPlayer::GetInstance().SetFrameWrittenCallback([] {
+    QueueOnObject(&Settings::Instance(), [] { Settings::Instance().TriggerFIFOFrameWritten(); });
+  });
+}
+
+FIFOWindow::~FIFOWindow()
+{
+  FifoPlayer::GetInstance().SetFileLoadedCallback(nullptr);
+  FifoPlayer::GetInstance().SetFrameWrittenCallback(nullptr);
+}
+
+void FIFOWindow::CreateWidgets()
+{
+  auto* layout = new QGridLayout;
+
+  // Info
+  auto* info_group = new QGroupBox(tr("File Info"));
+  auto* info_layout = new QHBoxLayout;
+
+  m_info_label = new QLabel;
+  info_layout->addWidget(m_info_label);
+  info_group->setLayout(info_layout);
+
+  info_group->setMinimumHeight(100);
```

Comment:
Why does `info_group` need a hardcoded minimum height?

Reply:
I use so that the size stays consistent and everything stays in the same place regardless of the info label content

Reply:
In that case, it should be something like `setFixedHeight(QFontMetrics(font()).lineSpacing() * 3);`